fix(langchain): Prevent streaming token leakage in middleware internal model calls

devbyteai opened the pull request:
## Summary

Fixes #34382

This PR fixes the streaming token leakage bug where internal model calls made by middleware (like `SummarizationMiddleware`, `LLMToolSelectorMiddleware`, and `ToolEmulatorMiddleware`) were contaminating the parent agent's output stream.

---

## Problem

When combining streaming agents with middleware that make internal model calls, the internal model tokens leak into the parent agent's stream output.

**Error Behavior:**
When using `agent.astream()`, users see tokens from internal middleware model calls mixed into their agent's response stream. This breaks the expected streaming behavior where only the main agent's responses should appear.

**Reproduction:**
```python
from langchain.agents import create_agent
from langchain.agents.middleware import SummarizationMiddleware

agent = create_agent(
    model,
    middleware=[SummarizationMiddleware(model="openai:gpt-4o", trigger=("messages", 10))],
)

# When streaming, summarization model tokens leak into the stream!
async for chunk in agent.astream({"messages": [("user", "Hello")]}):
    print(chunk)  # Shows both agent AND summarization model tokens
```

---

## Root Cause Analysis

The issue stems from how LangChain's `ensure_config()` function handles callback inheritance.

**The Flow:**
1. Agent is invoked with `astream()`, which sets up streaming callbacks in `var_child_runnable_config`
2. Middleware hooks (`before_model`, `wrap_model_call`, `wrap_tool_call`) execute
3. Middleware calls internal model via `self.model.invoke()` or `self.model.ainvoke()`
4. `ensure_config()` (in `langchain_core/runnables/config.py:199-249`) automatically inherits callbacks from `var_child_runnable_config`
5. Internal model call now has streaming callbacks → tokens leak to parent stream

**Code Path:**
```python
# langchain_core/runnables/config.py:129
var_child_runnable_config: ContextVar[RunnableConfig | None] = ContextVar(
    "child_runnable_config", default=None
)

# langchain_core/runnables/config.py:199-225
def ensure_config(config: RunnableConfig | None = None) -> RunnableConfig:
    empty = RunnableConfig(...)
    if var_config := var_child_runnable_config.get():
        empty.update(...)  # <-- Inherits callbacks from parent context!
    ...
```

**Affected Middlewares:**
1. `SummarizationMiddleware` - calls `self.model.invoke()` to generate summaries
2. `LLMToolSelectorMiddleware` - calls `structured_model.invoke()` to select tools
3. `ToolEmulatorMiddleware` - calls `self.model.invoke()` to emulate tool responses

---

## Solution

Pass `config={"callbacks": []}` to all internal model calls in affected middlewares. This explicitly clears the inherited streaming callbacks while preserving normal model execution.

**Changes in `libs/langchain_v1/langchain/agents/middleware/summarization.py`:**

```python
# Before (buggy):
response = self.model.invoke(self.summary_prompt.format(messages=trimmed_messages))

# After (fixed):
# Pass empty callbacks to prevent inheriting streaming callbacks from
# parent context, which would leak internal model output to agent stream
response = self.model.invoke(
    self.summary_prompt.format(messages=trimmed_messages),
    config={"callbacks": []},
)
```

**Changes in `libs/langchain_v1/langchain/agents/middleware/tool_selection.py`:**

```python
# Before (buggy):
response = structured_model.invoke([...])

# After (fixed):
# Pass empty callbacks to prevent inheriting streaming callbacks from
# parent context, which would leak internal model output to agent stream
response = structured_model.invoke([...], config={"callbacks": []})
```

**Changes in `libs/langchain_v1/langchain/agents/middleware/tool_emulator.py`:**

```python
# Before (buggy):
response = self.model.invoke([HumanMessage(prompt)])

# After (fixed):
# Pass empty callbacks to prevent inheriting streaming callbacks from
# parent context, which would leak internal model output to agent stream
response = self.model.invoke([HumanMessage(prompt)], config={"callbacks": []})
```

---

## Testing

All 43 tests for affected middlewares pass:

```
tests/unit_tests/agents/middleware/implementations/test_summarization.py

test_summarization_middleware_initialization PASSED
test_summarization_middleware_no_summarization_cases PASSED
test_summarization_middleware_helper_methods PASSED
test_summarization_middleware_summary_creation PASSED
test_summarization_middleware_trim_limit_none_keeps_all_messages PASSED
test_summarization_middleware_profile_inference_triggers_summary PASSED
test_summarization_middleware_token_retention_advances_past_tool_messages PASSED
test_summarization_middleware_missing_profile PASSED
test_summarization_middleware_full_workflow PASSED
test_summarization_middleware_full_workflow_async PASSED
test_summarization_middleware_keep_messages PASSED
test_summarization_middleware_validation_edge_cases (8 variants) PASSED
test_summarization_middleware_multiple_triggers PASSED
test_summarization_middleware_profile_edge_cases PASSED
test_summarization_middleware_trim_messages_error_fallback PASSED
test_summarization_middleware_binary_search_edge_cases PASSED
test_summarization_middleware_find_safe_cutoff_point PASSED
test_summarization_middleware_zero_and_negative_target_tokens PASSED
test_summarization_middleware_async_error_handling PASSED
test_summarization_middleware_cutoff_at_boundary PASSED
test_summarization_middleware_deprecated_parameters_with_defaults PASSED
test_summarization_middleware_fraction_trigger_with_no_profile PASSED
test_summarization_adjust_token_counts PASSED
test_summarization_middleware_many_parallel_tool_calls_safety PASSED
test_summarization_middleware_find_safe_cutoff_advances_past_tools PASSED
test_summarization_middleware_cutoff_at_start_of_tool_sequence PASSED

tests/unit_tests/agents/middleware/implementations/test_tool_selection.py

TestLLMToolSelectorBasic::test_sync_basic_selection PASSED
TestLLMToolSelectorBasic::test_async_basic_selection PASSED
TestMaxToolsLimiting::test_max_tools_limits_selection PASSED
TestMaxToolsLimiting::test_no_max_tools_uses_all_selected PASSED
TestAlwaysInclude::test_always_include_tools_present PASSED
TestAlwaysInclude::test_always_include_not_counted_against_max PASSED
TestAlwaysInclude::test_multiple_always_include_tools PASSED
TestDuplicateAndInvalidTools::test_duplicate_tool_selection_deduplicated PASSED
TestDuplicateAndInvalidTools::test_max_tools_with_duplicates PASSED
TestEdgeCases::test_empty_tools_list_raises_error PASSED

============================== 43 passed in 0.31s ==============================
```

Full middleware test suite (335 tests) also passes:
```
======================== 335 passed, 1 warning in 12.35s ========================
```

---

## Backward Compatibility

**No breaking changes** - This fix only affects internal middleware behavior:

1. **External API unchanged** - Middleware constructors and methods have the same signatures
2. **Agent streaming unchanged** - Users still receive agent responses via `astream()`
3. **Middleware execution unchanged** - Internal model calls still work correctly, just without leaking tokens

The only observable change is that internal middleware model calls no longer appear in the agent's stream - which is the correct and expected behavior.

---

## Design Decisions

**Why empty callbacks `[]` instead of `None`:**
Looking at `ensure_config()` implementation (line 226-236), it only updates config values when `v is not None`. Passing `callbacks=None` would not override inherited callbacks, but `callbacks=[]` explicitly sets an empty list.

**Why fix at middleware level instead of core:**
1. **Minimal scope** - Only affects specific methods that need isolation
2. **No core changes** - Doesn't modify `ensure_config()` behavior for all users
3. **Explicit intent** - Each internal call clearly documents why callbacks are cleared
4. **Backward compatible** - Existing behavior for all other code paths preserved

**Why fix all three middlewares:**
Comprehensive search found all middlewares with internal model calls. Fixing only one would leave the others with the same bug. As senior developers, we ensure the fix is complete.

---

## Future Considerations

1. **New middlewares** - Any future middleware with internal model calls should follow this same pattern
2. **Helper function** - If this pattern becomes common, consider adding `get_isolated_config()` utility
3. **Framework solution** - Could add automatic isolation at the middleware execution layer

---

## Related Issues

- [LangChain JS #9455](https://github.com/langchain-ai/langchainjs/issues/9455) - Same issue in JavaScript SDK (Summarization Middleware internal model.invoke() streamed to UI)

devbyteai committed:
fix(langchain): Prevent streaming token leakage in middleware internal model calls

When middleware like SummarizationMiddleware, LLMToolSelectorMiddleware, or
ToolEmulatorMiddleware makes internal model calls (e.g., for summarization or
tool selection), these calls were inheriting streaming callbacks from the parent
context via ensure_config(). This caused internal model tokens to leak into the
agent's output stream.

Fix: Pass config={"callbacks": []} to all internal model.invoke() and
model.ainvoke() calls to prevent inheriting parent streaming callbacks.

Fixes #34382

devbyteai committed:
style: Fix ruff formatting in tool_emulator.py

keenranger commented:
hi @devbyteai I wonder you have no problem for langsmith logging with this config. I Actually middleware model appeared in log, but w/o actually input/output or function called

devbyteai committed:
fix(middleware): preserve LangSmith tracing while blocking streaming callbacks

The previous fix used `config={"callbacks": []}` which blocked ALL callbacks
including LangSmith tracing. This caused middleware internal model calls to
appear in LangSmith logs but without input/output details.

This update introduces selective callback filtering via `get_internal_call_config()`:
- PRESERVES: BaseTracer subclasses (LangSmith, custom tracers)
- PRESERVES: UsageMetadataCallbackHandler (token tracking)
- BLOCKS: Streaming callbacks (prevents token leakage to parent stream)

Changes:
- Add _callback_utils.py with get_internal_call_config() utility
- Update SummarizationMiddleware to use filtered callbacks
- Update LLMToolSelectorMiddleware to use filtered callbacks
- Update LLMToolEmulator to use filtered callbacks
- Add test_callback_isolation.py with 11 tests for callback filtering

devbyteai committed:
fix: resolve lint errors in test_callback_isolation.py

devbyteai committed:
fix: move type-only imports to TYPE_CHECKING block

devbyteai committed:
fix: add noqa comment for intentional lazy import

devbyteai committed:
style: add trailing comma for lint

devbyteai committed:
style: fix formatting and mypy unreachable warning

devbyteai committed:
style: fix test file formatting

ccurme reviewed (commented):
Thanks for this. It's preferable IMO to tag these so that they can be filtered during the stream, in case an application does want to use it. This is done for summarization in https://github.com/langchain-ai/langchain/pull/34775.
